pkg/csi/hostpathimpl.go

```diff
@@ -0,0 +1,264 @@
+			AccessibleTopology: []*csi.Topology{{
+				Segments: map[string]string{
+					pkg.KubernetesNodeIdentityKey: nodeName,
+				},
+			}},
+		},
+	}
+	log.Infof("CreateVolume: create volume %s(size: %d) successfully", volumeID, req.GetCapacityRange().GetRequiredBytes())
+	return response, nil
+}
+
+func (impl *hostPathImpl) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	reqCtx := getValue(ctx, ctxKeyDeleteVolume)
+	volumeID := reqCtx.pv.Name
+	parameters := reqCtx.pv.Spec.CSI.VolumeAttributes
+	_, hostPath, err := impl.getVerifiedHostPath(volumeID, parameters)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "DeleteVolume: getVerifiedHostPath error: %s", err.Error())
+	}
+	nodeName := reqCtx.nodeName
+
+	conn, err := impl.cs.getNodeConn(nodeName)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "DeleteVolume: failed to connect to node %s: %s", nodeName, err.Error())
+	}
+	defer conn.Close()
+	if err := conn.CleanPath(ctx, hostPath, true); err != nil {
+		return nil, status.Errorf(codes.Internal, "DeleteVolume: failed to delete hostPath %s: %s", hostPath, err.Error())
+	}
+	log.Infof("DeleteVolume: delete HostPath volume(%s) successfully", volumeID)
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+func (impl *hostPathImpl) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (resp *csi.NodePublishVolumeResponse, err error) {
+	volumeID := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
+
+	basePath, volumeHostPath, err := impl.getVerifiedHostPath(volumeID, req.VolumeContext)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "NodePublishVolume: getVerifiedHostPath error: %s", err.Error())
+	}
+
+	// create volume directory on the host
+	if err := impl.ns.osTool.MkdirAll(volumeHostPath, 0777); err != nil {
+		return nil, status.Errorf(codes.Internal, "NodePublishVolume: failed to create directory %s: %s", volumeHostPath, err.Error())
+	}
+	defer func() {
+		// clean up volume directory when error
+		if err != nil {
+			_ = impl.ns.osTool.Remove(volumeHostPath)
+		}
+	}()
+
+	// set project quota
+	setQuotaDone := false
+	var setQuotaErr error
+	tags := tags(req.VolumeContext)
+	capacity, err := tags.GetInt64(volumeCapacityTag)
+	if err != nil {
+		log.Warningf("NodePublishVolume: failed to get %s, err: %s", volumeCapacityTag, err.Error())
+	} else {
+		setQuotaErr = impl.setProjectQuota(ctx, basePath, volumeHostPath, capacity)
+		if setQuotaErr != nil {
+			log.Warningf("NodePublishVolume: failed to set project quota, err: %s", setQuotaErr)
+		} else {
+			setQuotaDone = true
+		}
+	}
+	enforceCapacityLimit, _ := tags.GetBool(enforceCapacityLimitScTag)
+	if enforceCapacityLimit && !setQuotaDone {
+		return nil, status.Errorf(codes.Internal,
+			"NodePublishVolume: failed to set project quota, err: %s", setQuotaErr)
+	}
+
+	// mount target path
+	if err := impl.mountHostPathVolume(ctx, req, volumeHostPath); err != nil {
```

Comment:
err is shadowed, so if err at line 134 is not nil, may cause the volume successfully mounted but then be removed in line 125. Or fail to mount, but the volume directory is not removed.

Reply:
https://go.dev/play/p/LzFI_sdYmO8
<img width="513" alt="image" src="https://github.com/apecloud/open-local/assets/48559547/3706dc20-21e9-4cea-8651-902ef8eeefa1">
if return err of "mountHostPathVolume" is not nil, may cause the clean job in defer can not be triggered

Reply:
<img width="503" alt="image" src="https://github.com/apecloud/open-local/assets/48559547/f09784ac-4be8-4dc4-90c8-9b8b2a5bd494">

Reply:
Please be aware that the code you've provided is not identical to the code in the PR, which explicitly returns an error, so the correct err value can be seen within the defer statement.

![image](https://github.com/apecloud/open-local/assets/2639200/3e769836-2675-4f07-83c0-45e7e74adfdb)